src/Microsoft.FeatureManagement/Targeting/ContextualTargetingFeatureVariantAssigner.cs

```diff
@@ -0,0 +1,143 @@
+            }
+
+            if (targetingContext == null)
+            {
+                throw new ArgumentNullException(nameof(targetingContext));
+            }
+
+            FeatureDefinition featureDefinition = variantAssignmentContext.FeatureDefinition;
+
+            if (featureDefinition == null)
+            {
+                throw new ArgumentNullException(nameof(variantAssignmentContext.FeatureDefinition));
+            }
+
+            if (featureDefinition.Variants == null)
+            {
+                throw new ArgumentNullException(nameof(featureDefinition.Variants));
+            }
+
+            FeatureVariant variant = null;
+
+            FeatureVariant defaultVariant = null;
+
+            double cumulativePercentage = 0;
+
+            var cumulativeGroups = new Dictionary<string, double>();
+
+            foreach (FeatureVariant v in featureDefinition.Variants)
+            {
+                if (defaultVariant == null && v.Default)
+                {
+                    defaultVariant = v;
+                }
+
+                TargetingFilterSettings targetingSettings = v.AssignmentParameters.Get<TargetingFilterSettings>();
+
+                if (targetingSettings == null)
+                {
+                    if (v.Default)
+                    {
+                        //
+                        // Valid to omit audience for default variant
+                        continue;
+                    }
+                    else
+                    {
+                        targetingSettings = new TargetingFilterSettings();
+                    }
+                }
+
+                if (!TargetingEvaluator.TryValidateSettings(targetingSettings, out string paramName, out string reason))
+                {
+                    throw new ArgumentException(reason, paramName);
+                }
+
+                AccumulateAudience(targetingSettings.Audience, cumulativeGroups, ref cumulativePercentage);
+
+                if (TargetingEvaluator.IsTargeted(targetingSettings, targetingContext, _options.IgnoreCase, featureDefinition.Name))
+                {
+                    variant = v;
+
+                    break;
+                }
+            }
+
+            return new ValueTask<FeatureVariant>(variant);
+        }
+
+        /// <summary>
+        /// Accumulates percentages for groups and the default rollout for an audience.
+        /// </summary>
+        /// <param name="audience">The audience that will have its percentages updated based on currently accumulated percentages</param>
+        /// <param name="cumulativePercentage">The current cumulative default rollout percentage</param>
+        /// <param name="cumulativeGroups">The current cumulative rollout percentage for each group</param>
+        private static void AccumulateAudience(Audience audience, Dictionary<string, double> cumulativeGroups, ref double cumulativePercentage)
```

Comment:
>cumulativePercentage [](http://example.com/codeflow?start=122&length=20)

I found it easier to understand if we name it `cumulativeDefaultPercentage`.

Reply:
Updated.